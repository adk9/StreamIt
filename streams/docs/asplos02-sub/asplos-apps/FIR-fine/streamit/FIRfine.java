import streamit.StreamIt;
import streamit.Pipeline;
import streamit.SplitJoin;
import streamit.Filter;
import streamit.Channel;

public class FIRfine extends StreamIt
{
    public static void main (String [] args)
    {
        new FIRfine ().run (args);
    }
    
    public void init ()
    {
        add (new FloatSource (10000));
        add (new FIR (64));
	add (new FloatPrinter (10000));
    }
}

class FIR extends Pipeline
{
    FIR (int N)
    {

    }

    public void init (final int N)
    {
	int i;
	add(new Filter() {
		public void init() {
		    this.input = new Channel(Float.TYPE, 1, 1);
		    this.output = new Channel(Float.TYPE, 2);
		}
		public void work() {
		    this.output.pushFloat(this.input.popFloat());
		    this.output.pushFloat(0);
		}
	    });
	for(i=0; i<N; i++)
	    add (new SingleMultiply(i));
	add(new Filter() {
		public void init() {
		    this.input = new Channel(Float.TYPE, 2, 2);
		    this.output = new Channel(Float.TYPE, 1);
		}
		public void work() {
		    this.input.popFloat();
		    this.output.pushFloat(this.input.popFloat());
		}
	    });
    }
}

class SingleMultiply extends Filter
{
    SingleMultiply(int i)
    {

    }

    float W;
    public void init(final int i) {
	W = 2*i*i/((float)i+1);
	this.input = new Channel(Float.TYPE, 2, 2);
	this.output = new Channel(Float.TYPE, 2);
    }
    
    public void work() {
	float c, s; 
	c = this.input.popFloat();
	s = this.input.popFloat();
	this.output.pushFloat(c);
	this.output.pushFloat(s+c*W);
    }
}

class FloatSource extends Filter
{
    FloatSource (float maxNum)
    {

    }
    
    float num;
    float maxNum;
    
    public void init (float maxNum2)
    {
        output = new Channel (Float.TYPE, 1);
        this.maxNum = maxNum2;
        this.num = 0;
    }
    
    public void work ()
    {
        output.pushFloat (num);
        num++;
        if (num == maxNum) num = 0;
    }
}

class FloatPrinter extends Filter 
{
    FloatPrinter (int x)
    {

    }
    int x;
    public void init (int x2)
    {
        input = new Channel (Float.TYPE, 1);
        this.x = x2;

    }
    public void work ()
    {
      System.out.println(input.popFloat ());
    }
}

class FloatIdentity extends Filter
{
    public void init ()
    {
        input = new Channel (Float.TYPE, 1);
        output = new Channel (Float.TYPE, 1);
    }
    public void work ()
    {
        float i = input.popFloat ();
        output.pushFloat (i);
    }
}
