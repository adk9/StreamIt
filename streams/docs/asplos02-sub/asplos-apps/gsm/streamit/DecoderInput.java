// This is what used to be DecoderInput--now it will need to be
// inlined into every class that was using it.

//class DecoderInput
//{

//member variables!
public short[] mLarParameters = new short[8];
public short[] mLtpOffset = new short[4];
public short[] mLtpGain = new short[4];
public short[] mRpeGridPosition = new short[4];
public short[] mRpeMagnitude = new short[4];
public short[][] mSequence = new short[4][13];

public void getParameters(short[] input)
{
    int input_index = 0;
    int num_bits = 0;
    for(int i = 0; i < 8; i++)
	{
	
	    switch(i)
		{
		case 0:
		case 1:      num_bits = 6;
		    break;
		case 2:      
		case 3:      num_bits = 5;
		    break;
		case 4:      
		case 5:      num_bits = 4;
		    break;
		case 6:      
		case 7:      num_bits = 3;
		    break;
		}
	
	  
	    mLarParameters[i] = 0;
	    for (int j = 0; j < num_bits; j++, input_index++)
		{
		    mLarParameters[i] |= input[input_index] << (num_bits - 1 - i);
		}
	}
    
    //Sub-frames 1 through 4!
    for (int k = 0; k < 4; k++)
	{
	    mLtpOffset[k] = 0;
	    for (int l = 0; l < 7; l++)
		{
		    mLtpOffset[k] |= input[input_index] << (6 - l);
		    input_index++;
		}
	    mLtpGain[k] = 0;
	    for (int l = 0; l < 2; l++)
		{
		    mLtpGain[k] |= input[input_index] << (1 - l);
		    input_index++;
		}
	    mRpeGridPosition[k] = 0;
	    for (int l = 0; l < 2; l++)
		{
		    mRpeGridPosition[k] |= input[input_index] << (1 - l);
		    input_index++;
		}
	    mRpeMagnitude[k] = 0;
	    for (int l = 0; l < 6; l++)
		{
		    mRpeMagnitude[k] |= input[input_index] << (5 - l);
		    input_index++;
		}
	    for(int l = 0; l < 13; l++)
		{
		    mSequence[k][l] = 0;
		    for (int m = 0; m < 3; m++)
			{
			    mSequence[k][l] |= input[input_index] << (2 - m);
			    input_index++;
			}
	    
		}
	}
    //System.out.println(input_index);
    //System.out.println(input.length - input_index);
    
}

//}


