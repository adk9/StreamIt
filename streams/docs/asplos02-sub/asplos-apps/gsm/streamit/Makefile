all:
	# get the include files resolved
	cpp StGsmDecoder.java Gsm.java
	# eliminate the #'s generated by including
	perl -pi -e 's/#.*\n//g' Gsm.java
	# rename StGsmDecoder to Gsm
	perl -pi -e 's/StGsmDecoder/Gsm/g' Gsm.java
	# compile with jikes
	jikes Gsm.java

clean:
	rm -f *~
	rm -f *.class
	rm -f Gsm.java
	rm -f BinaryDecoderOutput1