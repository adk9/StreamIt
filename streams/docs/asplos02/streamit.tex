\section{The StreamIt Language}
\label{sec:streamit}

1 the streamit language

  - show software radio diagram
  - show some software radio code
  - give standard schpiel

