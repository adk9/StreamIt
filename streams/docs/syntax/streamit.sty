%
% streamit.sty: declare a `streamit' language for listings
% David Maze <dmaze@cag.lcs.mit.edu>
% $Id: streamit.sty,v 1.2 2002-08-21 19:07:13 dmaze Exp $
%

% Cause the `listings' package to exist, and define a `streamit'
% language for it...
\RequirePackage{listings}

% What should basicstyle be set to here?  Its default value doesn't
% cause inline code segments to be highlighted at all.  \sffamily
% gets around this, but computer modern sans serif bold is too heavy
% compared to the rest of the text (this seems odd to me).  \ttfamily
% doesn't have bold.  We might experiment with PS fonts, or with using
% \sffamily but also setting keywordstyle to empty.
%
% ...experimenting with postscript fonts: \usepackage{palatino} or
% \usepackage{times} gets us Helvetica for \sffamily, which is fine,
% but it looks bad and is just too big besides.  \ttfamily then is
% Courier, though, which has bold and arguably looks better than
% anything else we've looked at.  Courier bold is noticably heavier
% than not-bold, though.
%
% Combinations that seem to work reasonably:
% -- CM, basicstyle=\sffamily,keywordstyle={}
% -- CM, basicstyle=\ttfamily (no boldness)
% -- PS, basicstyle=\ttfamily
%
% The literate option also looks better if we're using
% \sffamily, but isn't strictly speaking correct.
\lstdefinelanguage{StreamIt}%
{keywords={filter,pipeline,splitjoin,int,float,complex,void,struct,%
init,work,push,pop,peek,feedbackloop,enqueue,add,loop,body,%
split,join,duplicate,roundrobin,template,string,continue,%
break,do,handler,for,while},%
comment=[l]//,comment=[s]{/*}{*/},%
%literate={->}{$\rightarrow$}1%
}
\lstset{language=StreamIt,flexiblecolumns=true,basicstyle=\ttfamily,%
  keywordstyle={},commentstyle=\slshape}

