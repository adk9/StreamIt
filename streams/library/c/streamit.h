#ifndef STREAMIT_H
#define STREAMIT_H

typedef enum stream_type {
  INVALID_STREAM_TYPE,
  FILTER,
  PIPELINE,
  SPLIT_JOIN,
  FEEDBACK_LOOP
} stream_type;
typedef enum splitjoin_type {
  ROUND_ROBIN,
  WEIGHTED_ROUND_ROBIN,
  DUPLICATE,
  COMBINE,
  NULL_SJ
} splitjoin_type;
typedef enum split_or_join {
  SPLITTER,
  JOINER
} split_or_join;
typedef enum in_or_out {
  INPUT,
  OUTPUT
} in_or_out;
typedef struct latency_list {
  int val;
  struct latency_list *next;
} latency_list;

typedef struct latency_range {
  int min_val;
  int max_val;
} latency_range;

typedef enum latency_special {
  _BEST_EFFORT
} latency_special;

typedef union latency {
  latency_list *list;
  latency_range range;
  latency_special special;
} latency;

#define LATENCY_BEST_EFFORT ((latency){ special: _BEST_EFFORT })
typedef void (*streamit_handler)(void *);
typedef void (*work_fn)(void *);
typedef void (*message_fn)(void *data, void *params);
typedef message_fn *interface_table;
typedef struct tape {
  void *data;
  int read_pos;
  int write_pos;
  int data_size;
  int tape_length;
} tape;
#define INCR_TAPE_WRITE(t) \
  ((++((t)->write_pos) >= (t)->tape_length) ? \
    (t)->write_pos = 0 : (t)->write_pos)
#define INCR_TAPE_READ(t) \
  ((++((t)->read_pos) >= (t)->tape_length) ? \
    (t)->read_pos = 0 : (t)->read_pos)
#define PUSH_TAPE(t, type, d) \
  (((type *)((t)->data))[INCR_TAPE_WRITE(t)] = (d))
#define PEEK_TAPE(t, type, n) \
  (((type *)(t)->data)[((t)->read_pos+n+1)%(t)->tape_length])
#define POP_TAPE(t, type) \
  (((type *)((t)->data))[INCR_TAPE_READ(t)])
#define COPY_TAPE_ITEM(s, d) \
  (memcpy((d)->data + (d)->write_pos * (d)->data_size, \
          (s)->data + (s)->read_pos * (s)->data_size, \
          (d)->data_size))
#define PUSH(c, type, d) PUSH_TAPE((c)->output_tape, type, d)
#define PEEK(c, type, n) PEEK_TAPE((c)->input_tape, type, n)
#define POP(c, type) POP_TAPE((c)->input_tape, type)
#define FEEDBACK_DELAY(d, c, n, t, f) { \
  int i; \
  for (i = 0; i < (n); i++) { \
    PUSH_TAPE((c)->type_data.splitjoin_data.joiner.tape[1], t, f((d), i)); \
  } \
}
typedef struct one_to_many {
  splitjoin_type type;
  int fan;
  int *ratio;
  int slots;
  tape *one_tape, **tape, **tcache;
} one_to_many;
struct stream_context;

typedef struct stream_context_list {
  struct stream_context *context;
  struct stream_context_list *next;
} stream_context_list;

typedef struct pipeline_type_data {
  stream_context_list *first_child;
  stream_context_list *last_child;
} pipeline_type_data;
typedef struct splitjoin_type_data {
  stream_context_list *first_child;
  stream_context_list *last_child;
  one_to_many splitter;
  one_to_many joiner;
} splitjoin_type_data;

typedef union stream_type_data {
  pipeline_type_data pipeline_data;
  splitjoin_type_data splitjoin_data;
} stream_type_data;
typedef struct stream_context {
  void *stream_data;
  stream_type type;
  int peek_size, pop_size, push_size;
  work_fn work_function;
  struct stream_context *parent;
  tape *input_tape;
  tape *output_tape;
  stream_type_data type_data;
} stream_context;
stream_context *create_context(void *p);
typedef struct portal_receiver {
  struct portal_receiver *next;
  stream_context *context;
  interface_table vtbl;
} portal_receiver;
typedef struct portal {
  portal_receiver *receiver;
} _portal, *portal;
void set_stream_type(stream_context *c, stream_type type);
void set_peek(stream_context *c, int peeks);
void set_pop(stream_context *c, int pops);
void set_push(stream_context *c, int pushes);
void set_work(stream_context *c, work_fn f);
void set_teardown(stream_context *c, streamit_handler f);
void register_child(stream_context *c, stream_context *child);
stream_context *get_parent(stream_context *c);
void create_tape(stream_context *a, stream_context *b,
                 int data_size, int tape_length);
void set_to_canon(stream_context *c, streamit_handler f);
void set_from_canon(stream_context *c, streamit_handler f);
void set_splitter(stream_context *c, splitjoin_type type, int n, ...);
void set_joiner(stream_context *c, splitjoin_type type, int n, ...);
void create_splitjoin_tape(stream_context *container,
                           split_or_join sj,
                           in_or_out io,
                           int slot,
                           stream_context *other,
                           int data_size, int tape_length);
void run_splitter(stream_context *c);
void run_joiner(stream_context *c);
portal create_portal(void);
void register_receiver(portal p, stream_context *receiver,
                       interface_table vtbl, latency l);
/* void register_sender(portal p, stream_context *sender, latency l); */
void send_message(portal p, int msgid, latency l, void *params);
void connect_tapes(stream_context *c);
void streamit_run(stream_context *c);

#endif /* STREAMIT_H */
