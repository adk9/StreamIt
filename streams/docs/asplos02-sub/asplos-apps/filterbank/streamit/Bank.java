/*
 *  Copyright 2002 Massachusetts Institute of Technology
 *
 *  Permission to use, copy, modify, distribute, and sell this software and its
 *  documentation for any purpose is hereby granted without fee, provided that
 *  the above copyright notice appear in all copies and that both that
 *  copyright notice and this permission notice appear in supporting
 *  documentation, and that the name of M.I.T. not be used in advertising or
 *  publicity pertaining to distribution of the software without specific,
 *  written prior permission.  M.I.T. makes no representations about the
 *  suitability of this software for any purpose.  It is provided "as is"
 *  without express or implied warranty.
 */


// This is the complete FIR pipeline

import streamit.*;
import streamit.io.*;

/**
 * Class FirFilter
 *
 * Implements an FIR Filter
 */

public class Bank extends StreamIt {
 

    
class source extends Filter {
    int N;
    float[] r;
    public source(float[] r) {super(r);}
    public void init(float[] r){
	output = new Channel(Float.TYPE,r.length);
	this.r=r;
	N=r.length;
    }
    public void work(){
	for(int i=0;i<N;i++)
	     output.pushFloat(r[i]);
    }
}

class sink extends Filter{
    int N;
    public sink(int N) {super(N);}
    public void init(int N){
	input = new Channel(Float.TYPE, N);
	this.N=N;
	//setPop(N);
    }
    public void work() {
	System.out.println("Starting");

	for (int i=0; i< N;i++)
	    {
		System.out.print("This is ");
		System.out.print(i);
		System.out.print(" : ");
		System.out.println(input.popFloat());
	    }
	    
    }
}


        
    static public void main(String[] t)
    {
	Bank test=new Bank();
	test.run(t);
    }


    /*   public Bank (int N,float[] H,float[] F)
    {
        //super (N,H,F);
	}*/

    public void init( /* int N,float[] H,float[] F */) {
	
	  float[] H={1,2,3};
          float[] F={2,1,3};
	  float[] r={1,2,3,4,1,2};
	  int N=2;

  
	add (new source(r));
	add (new FIR(H));
	add (new DownSamp(N));
	add (new UpSamp(N));
	add (new FIR(F));
	add (new sink(r.length));
    }
    

}























































