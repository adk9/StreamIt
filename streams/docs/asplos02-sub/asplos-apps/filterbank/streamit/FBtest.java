import streamit.*;
import streamit.io.*;

class source extends Filter {
    int N;
    float[] r;
    public source(float[] r) {super(r);}
    public void init(float[] r){
	output = new Channel(Float.TYPE,r.length);
	this.r=r;
	N=r.length;
    }
    public void work(){
	for(int i=0;i<N;i++)
	     output.pushFloat(r[i]);
    }

}

class sink extends Filter{
    int N;
    public sink(int N) {super(N);}
    public void init(int N){
	input = new Channel(Float.TYPE, N);
	this.N=N;
	//setPop(N);
    }
    public void work() {
	System.out.println("Starting");

	for (int i=0; i< N;i++)
	    {
		System.out.print("This is ");
		System.out.print(i);
		System.out.print(" : ");
		System.out.println(input.popFloat());
}

	    
    }

}

class FBtest extends StreamIt {
    
 
    static public void main(String[] t)
    {
	FBtest test=new FBtest();
	test.run(t);
    }
    
   

    public void init() {
	
	float[] r=new float[5];
	r[0]=0;
	r[1]=1;
	r[2]=2;
	r[3]=3;
	r[4]=4;
	
	
	int N_samp=2;
	int N_ch=2;
	int N_col=3;
	float [][] F={
	    {1,2,3},
	    {2,1,3}
	};
	
	float [][] H={
	    {1,0,2},
	    {5,1,1}
	};
	
	add (new source(r));
	add (new FilterBank(N_samp,N_ch,N_col,H,F));
	add (new sink(r.length));
    }
    
    
}

	















