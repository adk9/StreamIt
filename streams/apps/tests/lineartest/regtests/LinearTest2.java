import streamit.*;

/**
 * Simple StreamIt application that contains some linear filters
 * that we can hopefully identify with the linear analysis pass
 * that I am going to be writing.
 **/
public class LinearTest2 extends StreamIt {
    public static void main(String[] args) {
	StreamIt app = new LinearTest2();
	app.run(args);
    }
    public void init() {
	this.add(new Source());
	this.add(new FloatIdentity());
	this.add(new FloatFilter());
	this.add(new Sink());
    }
}


/**
 * Single, monotonically increasing source of floating point numbers.
 **/
class Source extends Filter {
    float INCREMENT = (float).01;
    float currentValue = 0;
    public void init() {
	output = new Channel(Float.TYPE, 1);
	this.currentValue = 0;
    }
    public void work() {
	output.pushFloat(this.currentValue);
	this.currentValue += INCREMENT;
    }
}

/**
 * Simple float sink that prints what it gets.
 **/
class Sink extends Filter {
    public void init() {
	input = new Channel(Float.TYPE, 1);
    }
    public void work() {
	System.out.println(input.popFloat());
    }
}

/**
 * About the simplest filter you can get that computes a linear function
 **/
class FloatIdentity extends Filter {
    public void init() {
	input = new Channel(Float.TYPE, 1);
	output = new Channel(Float.TYPE, 1);
    }
    public void work() {
	output.pushFloat(input.popFloat());
    }
}

/**
 * This is the filter which contains the work function that
 * I used to test the linear analysis.
 **/
class FloatFilter extends Filter {
    public void init() {
	input = new Channel(Float.TYPE, 3);
	output = new Channel(Float.TYPE, 3);
    }
    public void work() {
	// should result in 3rd col: [1,1,1] + 0
	output.pushFloat(input.peekFloat(0) +
			 input.peekFloat(1) +
			 input.peekFloat(2));

	// should result in 2nd col: [3,2,10] + 6
	output.pushFloat(1 +
			 2*input.peekFloat(1) +
			 input.peekFloat(2)*3 +
			 5 +
			 2*input.peekFloat(0)*5);

	// should result in 1st col: [0,0,0] + -11	
	output.pushFloat(-11);

	// pop off the values from the input tape
	input.popFloat();
	input.popFloat();
	input.popFloat();
    }
}
