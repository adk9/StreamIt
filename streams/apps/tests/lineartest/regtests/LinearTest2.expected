FloatFilter2_1.matrix
[[1.0+0.0i 4.0+0.0i 8.0+0.0i]
 [1.0+0.0i 3.0+0.0i 5.0+0.0i]
 [0.0+0.0i 2.0+0.0i 6.0+0.0i]
 [0.0+0.0i 0.0+0.0i 12.0+0.0i]]
FloatFilter2_1.vector
[[2.0+0.0i 7.0+0.0i 3.0+0.0i]]
FloatFilter3_1.matrix
[[0.2+0.0i]
 [0.6666666666666666+0.0i]
 [11.0+0.0i]
 [0.75+0.0i]]
FloatFilter3_1.vector
[[0.65625+0.0i]]
FloatFilter_1.matrix
[[0.0+0.0i 3.0+0.0i 1.0+0.0i]
 [0.0+0.0i 2.0+0.0i 1.0+0.0i]
 [0.0+0.0i 10.0+0.0i 1.0+0.0i]]
FloatFilter_1.vector
[[-11.0+0.0i 7.0+0.0i 0.0+0.0i]]
FloatIdentity_1.matrix
[[1.0+0.0i]]
FloatIdentity_1.vector
[[0.0+0.0i]]
